Specs/Scene/ModelExperimental/PickingPipelineStageSpec.js

```diff
@@ -0,0 +1,191 @@
+  ShaderBuilder,
+  Resource,
+  ResourceCache,
+} from "../../../Source/Cesium.js";
+import createScene from "../../createScene.js";
+import waitForLoaderProcess from "../../waitForLoaderProcess.js";
+
+describe("Scene/ModelExperimental/PickingPipelineStage", function () {
+  var boxVertexColors =
+    "./Data/Models/GltfLoader/BoxVertexColors/glTF/BoxVertexColors.gltf";
+  var boxInstanced =
+    "./Data/Models/GltfLoader/BoxInstanced/glTF/box-instanced.gltf";
+
+  var scene;
+  var gltfLoaders = [];
+
+  beforeAll(function () {
+    scene = createScene();
+  });
+
+  afterAll(function () {
+    scene.destroyForSpecs();
+  });
+
+  afterEach(function () {
+    var gltfLoadersLength = gltfLoaders.length;
+    for (var i = 0; i < gltfLoadersLength; ++i) {
+      var gltfLoader = gltfLoaders[i];
+      if (!gltfLoader.isDestroyed()) {
+        gltfLoader.destroy();
+      }
+    }
+    gltfLoaders.length = 0;
+    ResourceCache.clearForSpecs();
+  });
+
+  function getOptions(gltfPath, options) {
+    var resource = new Resource({
+      url: gltfPath,
+    });
+
+    return combine(options, {
+      gltfResource: resource,
+      incrementallyLoadTexture: false,
+    });
+  }
+
+  function loadGltf(gltfPath, options) {
+    var gltfLoader = new GltfLoader(getOptions(gltfPath, options));
+    gltfLoaders.push(gltfLoader);
+    gltfLoader.load();
+
+    return waitForLoaderProcess(gltfLoader, scene);
+  }
+
+  it("sets the picking variables in render resources", function () {
+    var renderResources = {
+      attributeIndex: 1,
+      pickId: undefined,
+      shaderBuilder: new ShaderBuilder(),
+      model: {
+        _resources: [],
+      },
+      runtimePrimitive: {
+        primitive: {},
+      },
+      runtimeNode: {
+        node: {},
+      },
+      uniformMap: {},
+    };
+
+    return loadGltf(boxVertexColors).then(function (gltfLoader) {
+      var components = gltfLoader.components;
+      var primitive = components.nodes[0].primitives[0];
+
+      var frameState = scene.frameState;
+      var context = frameState.context;
+      // Reset pick objects.
+      context._pickObjects = [];
+
+      PickingPipelineStage.process(renderResources, primitive, frameState);
+
+      var vertexDefineLines =
+        renderResources.shaderBuilder._vertexShaderParts.defineLines;
+      var fragmentDefineLines =
+        renderResources.shaderBuilder._fragmentShaderParts.defineLines;
+      var fragmentUniformLines =
+        renderResources.shaderBuilder._fragmentShaderParts.uniformLines;
+
+      expect(vertexDefineLines[0]).toEqual("ALLOWS_PICKING");
+      expect(fragmentDefineLines[0]).toEqual("ALLOWS_PICKING");
+      expect(fragmentUniformLines[0]).toEqual("uniform vec4 czm_pickColor;");
+
+      var pickObject = context._pickObjects["1"];
+      expect(pickObject).toBeDefined();
+      expect(pickObject.model).toBeDefined();
+      expect(pickObject.node).toBeDefined();
+      expect(pickObject.primitive).toBeDefined();
```

Comment:
Probably would be good to check that these point to exactly the model/node/primitive mentioned in the render resources above.